numpy/fft/pocketfft.c

```diff
@@ -0,0 +1,2408 @@
+/*
+ * This file is part of pocketfft.
+ * Licensed under a 3-clause BSD style license - see LICENSE.md
+ */
+
+/*
+ *  Main implementation file.
+ *
+ *  Copyright (C) 2004-2018 Max-Planck-Society
+ *  \author Martin Reinecke
+ */
+
+#include <math.h>
+#include <string.h>
+#include <stdlib.h>
+
+#include "npy_config.h"
+#define restrict NPY_RESTRICT
+
+#define RALLOC(type,num) \
+  ((type *)malloc((num)*sizeof(type)))
+#define DEALLOC(ptr) \
+  do { free(ptr); (ptr)=NULL; } while(0)
+
+#define SWAP(a,b,type) \
+  do { type tmp_=(a); (a)=(b); (b)=tmp_; } while(0)
+
+#ifdef __GNUC__
+#define NOINLINE __attribute__((noinline))
+#define WARN_UNUSED_RESULT __attribute__ ((warn_unused_result))
+#else
+#define NOINLINE
+#define WARN_UNUSED_RESULT
+#endif
+
+struct cfft_plan_i;
+typedef struct cfft_plan_i * cfft_plan;
+struct rfft_plan_i;
+typedef struct rfft_plan_i * rfft_plan;
+
+// adapted from https://stackoverflow.com/questions/42792939/
+// CAUTION: this function only works for arguments in the range [-0.25; 0.25]!
+static void my_sincosm1pi (double a, double *restrict res)
+  {
+  double s = a * a;
+  /* Approximate cos(pi*x)-1 for x in [-0.25,0.25] */
+  double r =     -1.0369917389758117e-4;
+  r = fma (r, s,  1.9294935641298806e-3);
+  r = fma (r, s, -2.5806887942825395e-2);
+  r = fma (r, s,  2.3533063028328211e-1);
+  r = fma (r, s, -1.3352627688538006e+0);
+  r = fma (r, s,  4.0587121264167623e+0);
+  r = fma (r, s, -4.9348022005446790e+0);
```

Comment:
We have to watch out here on Windows platforms (and possibly others). The Visual studio compiler will not treat `fma` as an intrinsic and will call an `fma` function instead, which performs a software emulation if it not provided the correct flags [[1]](https://stackoverflow.com/questions/34265982/automatically-generate-fma-instructions-in-msvc). However this software emulation is apparently really bad and does not provide the actual fma operation with a single rounding. This is the reason fma is not available in Python yet [[2]](https://bugs.python.org/issue29282).

Unfortunately, I do not have a Windows platform at hand where I could test the accuracy of this function. But we should probably do so to make sure the fma operations are compiled correctly. Also I apologize that I brought the `sincospi` up and it now turns out to be such a hassle to implement on different platforms. Maybe we should stick with the libm implementation of sin and cos.

Reply:
Also, libm `cos` wouldn't give me the `cos(x)-1` value accurately, which I need for the twiddle factors. Overall I think we will be doing well with the current implementation.

Reply:
`cos(x) - 1 = sin(x)**2/(cos(x) + 1)` should have good accuracy for `cos(x) > 0`.

Reply:
Thanks, that's very handy!
Still, the accuracy plots seem to indicate that we won't need to go there.